### PR TITLE
@types/got: Added create + generic body for GotFn

### PR DESCRIPTION
**Pull request by sebald**
- Added `got.create` to typings. This enables this https://github.com/sindresorhus/got/blob/master/advanced-creation.md

- Added optional (generic) return type for the `body` (see tests).

### CODE REVIEW

**Comment by sebald** [comment]
Hm, the linter doesn't allow something like this (`no-unnecessary-generics`):

```ts
<T extends Buffer | string | object = string>(url: GotUrl): GotPromise<T>;
```

But how are we supposed to allow setting a return type via generics?

**Review by ikokostya — CHANGES_REQUESTED** [review]

File: types/got/got-tests.ts [hunk 1]
```diff
@@ -13,37 +13,37 @@ let str: string;
 let buf: Buffer;
 
 got('todomvc.com')
-    .then(response => {
-        str = response.body;
-    })
-    .catch((error: got.GotError) => {
-        console.log(error.response.body);
-    });
+	.then(response => {
```
[the remark this answers is not in the document]
Comment (sebald) [reply types/got/got-tests.ts]: fixed

**Comment by LinusU** [comment]
I'm personally not a fan of adding a generic type for the body, since there is no enforcing going on anywhere. Actually, I think that `unknown` would be the correct type for the `body` property on response.

The reason for my thinking is that it will encourage people to not check what they get back, and then easily end up with the very same runtime errors that TypeScript was supposed to protect us from.

If the user wants to skip the compile-time checks, that's easy to do with an explicit `body as Foo`.

Just my 2¢ ☺️

**Comment by sebald** [comment]
@LinusU By default the typings behave the same. Adding the generics is optional. This way the typings are more generous. This is especially helpful if you're using `create`.

While this works:

```ts
const response = await got('sindresorhus.com', {json: true});
const body = response.body as { foo: string }; 
```

This is kinda cumbersome:

```ts
const api = got.create({
  options: got.mergeOptions(got.defaults.options, {
    json: true
  })
});

const response = await api.get('sindresorhus.com');

// Error, because `got.Response<string>` is inferred as `body`.
const body = response.body as { foo: string };

// Works
const body = response.body as any as { foo: string };
```

---

`unkown` maybe also an option, but I like sensible defaults. Also, if you share your (e.g. swagger) API typings and depending on the project size, you might want to skip the compile time check 🙂

**Comment by LinusU** [comment]
> but I like sensible defaults.

Me too, but in my opinion, it's not sensible to have an API that looks like it returns data of a specific type when in reality it can return anything. That's precisely why the `unknown` keyword was added...

The problem you are describing here:

> ```
> // Error, because `got.Response<string>` is inferred as `body`.
> ```

Seems to be another problem entirely, it's just that you don't notice it when using generics. Even if we merge this as-is that problem will still affect anyone not using the optional generics, or am I missing something?

Seems like we should carry thru the infered type of `options` when passed to create, probably thru a generics-parameter. That way we should be able to overload on `json: true` and return the correct body 🤔

**Comment by sebald** [comment]
> Seems to be another problem entirely, it's just that you don't notice it when using generics. Even if we merge this as-is that problem will still affect anyone not using the optional generics, or am I missing something?

IMHO the generic solves this, because you can set the type accordingly. E.g. if you don't trust your endpoint you can set it to `object` for JSONs.

> Seems like we should carry thru the infered type of options when passed to create, probably thru a generics-parameter. That way we should be able to overload on json: true and return the correct body 🤔

Inferring this might be tricky and error-prone, because of `got.mergeOptions`. Types may be widened.

---

The `unkown` type would be an option, but currently the target TS version is 2.3
`tslint` didn't allow me to use spread syntax in the `got.mergeOptions` because of this for example 😕

Passing it throw would mean adding a bunch of additional types, because the `settings` for `got.created` are a bit different from the `GotOptions`. Maybe it would be easier to make `create` generic to allow for the `body` type to be passed down to `GotFn`? 

For example:

```ts
interface GotObject<T extends Buffer | string | object = string> extends GotFn<T>, Record<'get' | 'post' | 'put' | 'patch' | 'head' | 'delete', got.GotFn<T>>{
    defaults: got.GotDefaults;
    stream: got.GotStreamFn & Record<'get' | 'post' | 'put' | 'patch' | 'head' | 'delete', got.GotStreamFn>;
    create: got.GotCreateFn;
    mergeOptions: got.GotMergeOptionsFn;

    RequestError: typeof RequestError;
    ReadError: typeof ReadError;
    ParseError: typeof ParseError;
    HTTPError: typeof HTTPError;
    MaxRedirectsError: typeof MaxRedirectsError;
    UnsupportedProtocolError: typeof UnsupportedProtocolError;
    CancelError: typeof CancelError;
    TimeoutError: typeof TimeoutError;
}

type GotCreateFn = <T extends Buffer | string | object = string, S extends GotOptions<string | null> = GotOptions<null>>(settings?: GotCreateSettings<S>) => GotObject<T>;
```

This would allow for 

```ts
const jsonApi = got.create<object>(got.mergeOptions(
    settings, { options: { json: true }})
);
jsonApi.get('http://todomvc.com?t=json')
    .then(r => {
        r.body; // $ExpectType object
    });
```

You still would have to assert the `body`, but would not need an additional `as any` at least.

**Comment by sebald** [comment]
I started using the typings with a generic `GotFn` and it still includes additional (unnecessary) type assertions. It also leaks `got.Response` typings in the application (it would be better if the rest of the applications doesn't have to know about this).

I am closing this PR, because the initial proposal seems not to conform with the original authors of the typings (which is absolutely fine!).

**Pull request closed without merging**